src/v/cluster/rm_stm.cc

```diff
@@ -1990,6 +1990,15 @@ void rm_stm::reconcile_mem_state() {
 }
 
 ss::future<> rm_stm::checkpoint_in_memory_state() {
+    if (unlikely(_mem_state.all_txns_count() > checkpoint_txns_limit)) {
```

Comment:
We either need to rename the method to `maybe_checkpoint_in_memory_state` or move the `all_txns_count` check to transfer_leadership. I prefer to move the check.